### imports

In [1]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [ ]:
import os
from tqdm import tqdm
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim

from torch.utils.data import DataLoader
from losses.final_loss import InverseDistanceBoundaryDiceLoss
from models.cbam_nnunet import nnUNet3D_CBAM
from utils.unets_helper_functions import (
    set_seed,
    save_checkpoint,
    final_PatchDataset_cbam,
    evaluate_full_volume_cbam,
    compute_per_class_dice
    
)

### data loading

In [3]:
set_seed(42)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)


Using device: cuda


In [4]:
FOLD = 0

with open(f"../data/splits/fold_{FOLD}/train.txt") as f:
    train_cases = f.read().splitlines()

with open(f"../data/splits/fold_{FOLD}/val.txt") as f:
    val_cases = f.read().splitlines()

print("Train cases:", len(train_cases))
print("Val cases:", len(val_cases))


Train cases: 33
Val cases: 9


In [5]:
PATCH_SIZE = 96
PATCHES_PER_CASE = 12
FOREGROUND_PROB = 0.7
NUM_WORKERS = 2
train_dataset = final_PatchDataset_cbam(train_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = PATCHES_PER_CASE,
                        patch_size = PATCH_SIZE,
                        foreground_prob = FOREGROUND_PROB,
                        augment=True)

val_dataset = final_PatchDataset_cbam(val_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = 1,
                        patch_size = PATCH_SIZE,
                        foreground_prob = FOREGROUND_PROB,
                        augment=False)


train_loader = DataLoader(
    train_dataset,
    batch_size=2,
    shuffle=True,
    num_workers=NUM_WORKERS,      
    pin_memory=True    
)

val_loader = DataLoader(
    val_dataset,
    batch_size=1,
    num_workers=NUM_WORKERS,    
    pin_memory=True
)

In [ ]:
model = nnUNet3D_CBAM(
    in_channels=1,
    out_channels=7,
    base_filters=24  
).to(device)

print("cbam nnU-Net style model Initialized")


cbam nnU-Net style model Initialized


In [7]:
EPOCHS = 100

optimizer = torch.optim.AdamW(model.parameters(), lr=2e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=EPOCHS
)

weights = torch.tensor([0.05, 1, 1, 1.2, 1.8, 1.8, 1]).to(device)

loss_fn = InverseDistanceBoundaryDiceLoss(
    epsilon=1e-5,
    lambda_weight=0.4,
    class_weights = weights
)

scaler = torch.cuda.amp.GradScaler()

C:\Users\sajju\AppData\Local\Temp\ipykernel_8864\4270202072.py:16: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [11]:
def train_one_epoch_cbam(model, loader, optimizer, scaler, loss_fn, device):

    model.train()
    total_loss = 0

    for images, labels, dist_maps in tqdm(loader):

        images = images.to(device)
        labels = labels.long().to(device)
        dist_maps = dist_maps.to(device)

        optimizer.zero_grad()

        with torch.amp.autocast("cuda"):

            outputs = model(images)

            # Deep supervision unpack
            out, ds2, ds3, ds4 = outputs


            loss_main = loss_fn(out, labels, dist_maps)
            loss_ds2  = loss_fn(ds2, labels, dist_maps)
            loss_ds3  = loss_fn(ds3, labels, dist_maps)
            loss_ds4  = loss_fn(ds4, labels, dist_maps)

            loss = (
                1.0 * loss_main +
                0.5 * loss_ds2 +
                0.25 * loss_ds3 +
                0.125 * loss_ds4
            )

        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()

        total_loss += loss.item()

    return total_loss / len(loader)

def validate_one_epoch_cbam(model, loader, loss_fn, device, num_classes=7):

    model.eval()
    total_loss = 0
    all_dices = []

    with torch.no_grad():
        for images, labels, dist_maps in loader:

            images = images.to(device)
            labels = labels.long().to(device)
            dist_maps = dist_maps.to(device)

            with torch.amp.autocast("cuda"):

                outputs = model(images)
                out = outputs[0]

                loss = loss_fn(out, labels, dist_maps)

            total_loss += loss.item()

            batch_dice = compute_per_class_dice(out, labels, num_classes)
            all_dices.append(batch_dice)

    mean_loss = total_loss / len(loader)

    all_dices = np.array(all_dices)
    mean_dices = np.mean(all_dices, axis=0)

    return mean_loss, mean_dices

#### weight - ([0.05, 1, 1, 1.2, 1.8, 1.8, 1])

In [ ]:
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "custom_loss_cbam")

os.makedirs(SAVE_DIR, exist_ok=True)

best_val_loss = float("inf")
best_dice = 0.0
best_parotid = 0.0
best_combined = 0.0

for epoch in range(EPOCHS):

    # ------------------ TRAIN ------------------
    train_loss = train_one_epoch_cbam(
        model,
        train_loader,
        optimizer,
        scaler,
        loss_fn,
        device
    )

    # ------------------ VALIDATE ------------------
    val_loss, val_dice = validate_one_epoch_cbam(
        model,
        val_loader,
        loss_fn,
        device
    )

    # Ignore background (class 0)
    mean_dice = val_dice[1:].mean()

    # Parotid (L=4, R=5)
    parotid_score = (val_dice[3] + val_dice[4]) / 2
    # 🔥 Combined metric (IMPORTANT)
    combined_score = 0.7 * mean_dice + 0.3 * parotid_score

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")
    print(f"Mean Dice: {mean_dice:.4f}")
    print(f"Parotid Dice (L,R): {val_dice[3]:.4f}, {val_dice[4]:.4f}")
    print(f"Combined Score: {combined_score:.4f}")

    # ------------------ SAVE BEST LOSS ------------------
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "04_iw_best_loss_model.pth")
        )
        print("Saved Best Loss Model")

    # ------------------ SAVE BEST DICE ------------------
    if mean_dice > best_dice:
        best_dice = mean_dice
        save_checkpoint(
            model,
            optimizer,
            epoch,
            mean_dice,
            os.path.join(SAVE_DIR, "04_iw_best_dice_model.pth")
        )
        print("Saved Best Dice Model")

    # ------------------ SAVE BEST PAROTID ------------------
    if parotid_score > best_parotid:
        best_parotid = parotid_score
        save_checkpoint(
            model,
            optimizer,
            epoch,
            parotid_score,
            os.path.join(SAVE_DIR, "04_iw_best_parotid_model.pth")
        )
        print("Saved Best Parotid Model")

    # ------------------ SAVE BEST COMBINED ------------------
    if combined_score > best_combined:
        best_combined = combined_score
        save_checkpoint(
            model,
            optimizer,
            epoch,
            combined_score,
            os.path.join(SAVE_DIR, "04_iw_best_combined_model.pth")
        )
        print("Saved Best Combined Model")

    # ------------------ SAVE LAST MODEL ------------------
    save_checkpoint(
        model,
        optimizer,
        epoch,
        val_loss,
        os.path.join(SAVE_DIR, "04_iw_last_model.pth")
    )

    # ------------------ LR PRINT ------------------
    lr = optimizer.param_groups[0]['lr']
    print(f"LR: {lr:.6f}")

    # ------------------ SCHEDULER ------------------
    scheduler.step()

NameError: name 'train_one_epoch_cbam' is not defined

In [10]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = nnUNet3D_CBAM(
    in_channels=1,
    out_channels=7,
    base_filters=24   # or try 32
).to(device)

model_path = "../experiments/final_cbam_fold0/best_combined_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


C:\Users\sajju\AppData\Local\Temp\ipykernel_8864\3396178766.py:10: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


Model loaded for testing.


In [11]:
mean_dice, std_dice = evaluate_full_volume_cbam(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Volume: 272x357x357 | Patches: 245 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.9221246965498694), np.float64(0.860959124669328), np.float64(0.7917169977709801), np.float64(0.8491824442242013), np.float64(0.835006891282726), np.float64(0.9043543123721484)]
Volume: 243x383x383 | Patches: 245 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.9149963334407858), np.float64(0.7965223502269936), np.float64(0.7197753996845492), np.float64(0.8137167579757496), np.float64(0.8540613004362554), np.float64(0.8933584892781475)]
Volume: 264x333x333 | Patches: 180 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8701675978068101), np.float64(0.7912959069800227), np.float64(0.7464767392840177), np.float64(0.7878436209560864), np.float64(0.793602437323989), np.float64(0.8811306441701358)]
Volume: 248x395x395 | Patche

In [13]:
model = nnUNet3D_CBAM(
    in_channels=1,
    out_channels=7,
    base_filters=24   # or try 32
).to(device)

print("cbam nnU-Net style model Initialized")


cbam nnU-Net style model Initialized


### revised weights - ([0.05, 1, 1.2, 1.5, 2, 2, 1])

In [15]:
EPOCHS = 40

optimizer = torch.optim.AdamW(model.parameters(), lr=2e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=EPOCHS
)

weights = torch.tensor([0.05, 1, 1.2, 1.5, 2, 2, 1]).to(device)

loss_fn = InverseDistanceBoundaryDiceLoss(
    epsilon=1e-5,
    lambda_weight=0.4,
    class_weights = weights
)

scaler = torch.cuda.amp.GradScaler()

C:\Users\sajju\AppData\Local\Temp\ipykernel_8864\1239421176.py:16: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [ ]:
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "custom_loss_cbam")

os.makedirs(SAVE_DIR, exist_ok=True)

best_val_loss = float("inf")
best_dice = 0.0
best_parotid = 0.0
best_combined = 0.0

for epoch in range(EPOCHS):

    # ------------------ TRAIN ------------------
    train_loss = train_one_epoch_cbam(
        model,
        train_loader,
        optimizer,
        scaler,
        loss_fn,
        device
    )

    # ------------------ VALIDATE ------------------
    val_loss, val_dice = validate_one_epoch_cbam(
        model,
        val_loader,
        loss_fn,
        device
    )

    # Ignore background (class 0)
    mean_dice = val_dice[1:].mean()

    # Parotid (L=4, R=5)
    parotid_score = (val_dice[3] + val_dice[4]) / 2
    # 🔥 Combined metric (IMPORTANT)
    combined_score = 0.7 * mean_dice + 0.3 * parotid_score

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")
    print(f"Mean Dice: {mean_dice:.4f}")
    print(f"Parotid Dice (L,R): {val_dice[3]:.4f}, {val_dice[4]:.4f}")
    print(f"Combined Score: {combined_score:.4f}")

    # ------------------ SAVE BEST LOSS ------------------
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "04_rw_best_loss_model.pth")
        )
        print("Saved Best Loss Model")

    # ------------------ SAVE BEST DICE ------------------
    if mean_dice > best_dice:
        best_dice = mean_dice
        save_checkpoint(
            model,
            optimizer,
            epoch,
            mean_dice,
            os.path.join(SAVE_DIR, "04_rw_best_dice_model.pth")
        )
        print("Saved Best Dice Model")

    # ------------------ SAVE BEST PAROTID ------------------
    if parotid_score > best_parotid:
        best_parotid = parotid_score
        save_checkpoint(
            model,
            optimizer,
            epoch,
            parotid_score,
            os.path.join(SAVE_DIR, "04_rw_best_parotid_model.pth")
        )
        print("Saved Best Parotid Model")

    # ------------------ SAVE BEST COMBINED ------------------
    if combined_score > best_combined:
        best_combined = combined_score
        save_checkpoint(
            model,
            optimizer,
            epoch,
            combined_score,
            os.path.join(SAVE_DIR, "04_rw_best_combined_model.pth")
        )
        print("Saved Best Combined Model")

    # ------------------ SAVE LAST MODEL ------------------
    save_checkpoint(
        model,
        optimizer,
        epoch,
        val_loss,
        os.path.join(SAVE_DIR, "04_rw_last_model.pth")
    )

    # ------------------ LR PRINT ------------------
    lr = optimizer.param_groups[0]['lr']
    print(f"LR: {lr:.6f}")

    # ------------------ SCHEDULER ------------------
    scheduler.step()

100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 0
Train Loss: 0.7235
Val Loss: 0.3127
Per Class Dice: [4.58449692e-01 2.22222265e-01 1.11112232e-01 3.33333335e-01
 5.56831040e-07 4.14338910e-01]
Mean Dice: 0.2162
Parotid Dice (L,R): 0.3333, 0.0000
Combined Score: 0.2013
Saved Best Loss Model
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000200


100%|██████████| 198/198 [05:50<00:00,  1.77s/it]



Epoch 1
Train Loss: 0.4732
Val Loss: 0.1931
Per Class Dice: [0.65689332 0.79878495 0.56670543 0.38501603 0.62877414 0.88347527]
Mean Dice: 0.6526
Parotid Dice (L,R): 0.3850, 0.6288
Combined Score: 0.6089
Saved Best Loss Model
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000200


100%|██████████| 198/198 [06:09<00:00,  1.87s/it]



Epoch 2
Train Loss: 0.3267
Val Loss: 0.2031
Per Class Dice: [0.51873717 0.78291411 0.66057815 0.69272789 0.69884384 0.5868949 ]
Mean Dice: 0.6844
Parotid Dice (L,R): 0.6927, 0.6988
Combined Score: 0.6878
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000199


100%|██████████| 198/198 [05:46<00:00,  1.75s/it]



Epoch 3
Train Loss: 0.2862
Val Loss: 0.1146
Per Class Dice: [0.63606403 0.79099235 0.76944737 0.66175366 0.39966504 0.72336854]
Mean Dice: 0.6690
Parotid Dice (L,R): 0.6618, 0.3997
Combined Score: 0.6275
Saved Best Loss Model
LR: 0.000197


100%|██████████| 198/198 [06:13<00:00,  1.89s/it]



Epoch 4
Train Loss: 0.2737
Val Loss: 0.0569
Per Class Dice: [0.58777266 0.75105835 0.73432719 0.76302375 0.49001022 0.79191015]
Mean Dice: 0.7061
Parotid Dice (L,R): 0.7630, 0.4900
Combined Score: 0.6822
Saved Best Loss Model
Saved Best Dice Model
LR: 0.000195


100%|██████████| 198/198 [06:34<00:00,  1.99s/it]



Epoch 5
Train Loss: 0.2682
Val Loss: 0.0965
Per Class Dice: [0.72899364 0.63566452 0.81524593 0.62765385 0.74558362 0.78318359]
Mean Dice: 0.7215
Parotid Dice (L,R): 0.6277, 0.7456
Combined Score: 0.7110
Saved Best Dice Model
Saved Best Combined Model
LR: 0.000192


100%|██████████| 198/198 [05:48<00:00,  1.76s/it]



Epoch 6
Train Loss: 0.2492
Val Loss: 0.1741
Per Class Dice: [0.46834581 0.79261827 0.79226078 0.36597044 0.80585607 0.40493548]
Mean Dice: 0.6323
Parotid Dice (L,R): 0.3660, 0.8059
Combined Score: 0.6184
LR: 0.000189


100%|██████████| 198/198 [05:57<00:00,  1.81s/it]



Epoch 7
Train Loss: 0.2626
Val Loss: 0.1708
Per Class Dice: [0.63126496 0.83590086 0.8447138  0.77034761 0.67690232 0.66880783]
Mean Dice: 0.7593
Parotid Dice (L,R): 0.7703, 0.6769
Combined Score: 0.7486
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000185


100%|██████████| 198/198 [05:58<00:00,  1.81s/it]



Epoch 8
Train Loss: 0.2483
Val Loss: 0.0783
Per Class Dice: [0.77415266 0.71985853 0.75043558 0.60613946 0.53463932 0.70248051]
Mean Dice: 0.6627
Parotid Dice (L,R): 0.6061, 0.5346
Combined Score: 0.6350
LR: 0.000181


100%|██████████| 198/198 [05:59<00:00,  1.82s/it]



Epoch 9
Train Loss: 0.2432
Val Loss: 0.1019
Per Class Dice: [0.80377879 0.71531359 0.80735272 0.54422419 0.65606919 0.76499621]
Mean Dice: 0.6976
Parotid Dice (L,R): 0.5442, 0.6561
Combined Score: 0.6684
LR: 0.000176


100%|██████████| 198/198 [06:00<00:00,  1.82s/it]



Epoch 10
Train Loss: 0.2339
Val Loss: 0.0822
Per Class Dice: [0.59363966 0.80245974 0.61142498 0.78671622 0.67636082 0.80260502]
Mean Dice: 0.7359
Parotid Dice (L,R): 0.7867, 0.6764
Combined Score: 0.7346
Saved Best Parotid Model
LR: 0.000171


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 11
Train Loss: 0.2225
Val Loss: 0.0476
Per Class Dice: [0.51713704 0.83170473 0.94036871 0.6582745  0.86361751 0.91247459]
Mean Dice: 0.8413
Parotid Dice (L,R): 0.6583, 0.8636
Combined Score: 0.8172
Saved Best Loss Model
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000165


100%|██████████| 198/198 [05:52<00:00,  1.78s/it]



Epoch 12
Train Loss: 0.2292
Val Loss: 0.0607
Per Class Dice: [0.93073269 0.87177734 0.85549798 0.88870855 0.79659618 0.89202202]
Mean Dice: 0.8609
Parotid Dice (L,R): 0.8887, 0.7966
Combined Score: 0.8554
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000159


100%|██████████| 198/198 [06:02<00:00,  1.83s/it]



Epoch 13
Train Loss: 0.2176
Val Loss: 0.0821
Per Class Dice: [0.79993759 0.71157737 0.83446345 0.81976165 0.77876169 0.86522517]
Mean Dice: 0.8020
Parotid Dice (L,R): 0.8198, 0.7788
Combined Score: 0.8011
LR: 0.000152


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 14
Train Loss: 0.2150
Val Loss: 0.0850
Per Class Dice: [0.78944117 0.75664316 0.75944021 0.72146498 0.51275834 0.80077653]
Mean Dice: 0.7102
Parotid Dice (L,R): 0.7215, 0.5128
Combined Score: 0.6823
LR: 0.000145


100%|██████████| 198/198 [05:46<00:00,  1.75s/it]



Epoch 15
Train Loss: 0.2170
Val Loss: 0.0724
Per Class Dice: [0.9268127  0.84631752 0.63565619 0.86733075 0.67285641 0.78053741]
Mean Dice: 0.7605
Parotid Dice (L,R): 0.8673, 0.6729
Combined Score: 0.7634
LR: 0.000138


100%|██████████| 198/198 [05:44<00:00,  1.74s/it]



Epoch 16
Train Loss: 0.2221
Val Loss: 0.0656
Per Class Dice: [0.81349242 0.87548606 0.89541499 0.72799822 0.91187857 0.86706088]
Mean Dice: 0.8556
Parotid Dice (L,R): 0.7280, 0.9119
Combined Score: 0.8449
LR: 0.000131


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 17
Train Loss: 0.2214
Val Loss: 0.0719
Per Class Dice: [0.89885198 0.8200657  0.81518318 0.87950601 0.8852133  0.90447109]
Mean Dice: 0.8609
Parotid Dice (L,R): 0.8795, 0.8852
Combined Score: 0.8673
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000123


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 18
Train Loss: 0.2089
Val Loss: 0.1244
Per Class Dice: [0.80960979 0.83116294 0.66037146 0.5908963  0.66784055 0.77813437]
Mean Dice: 0.7057
Parotid Dice (L,R): 0.5909, 0.6678
Combined Score: 0.6828
LR: 0.000116


100%|██████████| 198/198 [06:01<00:00,  1.83s/it]



Epoch 19
Train Loss: 0.2044
Val Loss: 0.0723
Per Class Dice: [0.9238352  0.69503338 0.85016227 0.85605688 0.74195829 0.86115611]
Mean Dice: 0.8009
Parotid Dice (L,R): 0.8561, 0.7420
Combined Score: 0.8003
LR: 0.000108


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 20
Train Loss: 0.1966
Val Loss: 0.1025
Per Class Dice: [0.82226595 0.85295093 0.85870045 0.81736434 0.64129734 0.78090498]
Mean Dice: 0.7902
Parotid Dice (L,R): 0.8174, 0.6413
Combined Score: 0.7720
LR: 0.000100


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 21
Train Loss: 0.2171
Val Loss: 0.0372
Per Class Dice: [0.84835221 0.90636924 0.90789639 0.74207473 0.94666974 0.84368942]
Mean Dice: 0.8693
Parotid Dice (L,R): 0.7421, 0.9467
Combined Score: 0.8618
Saved Best Loss Model
Saved Best Dice Model
LR: 0.000092


100%|██████████| 198/198 [05:57<00:00,  1.81s/it]



Epoch 22
Train Loss: 0.1952
Val Loss: 0.0647
Per Class Dice: [0.92739285 0.87476061 0.75378429 0.89749807 0.73089396 0.91944431]
Mean Dice: 0.8353
Parotid Dice (L,R): 0.8975, 0.7309
Combined Score: 0.8290
LR: 0.000084


100%|██████████| 198/198 [05:58<00:00,  1.81s/it]



Epoch 23
Train Loss: 0.1946
Val Loss: 0.0421
Per Class Dice: [0.83284035 0.89488365 0.87846598 0.89885719 0.95256889 0.91943128]
Mean Dice: 0.9088
Parotid Dice (L,R): 0.8989, 0.9526
Combined Score: 0.9139
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000077


100%|██████████| 198/198 [05:57<00:00,  1.81s/it]



Epoch 24
Train Loss: 0.1950
Val Loss: 0.0835
Per Class Dice: [0.75726202 0.87458397 0.60624004 0.79928841 0.89117414 0.74180421]
Mean Dice: 0.7826
Parotid Dice (L,R): 0.7993, 0.8912
Combined Score: 0.8014
LR: 0.000069


100%|██████████| 198/198 [05:59<00:00,  1.82s/it]



Epoch 25
Train Loss: 0.2054
Val Loss: 0.0714
Per Class Dice: [0.92768187 0.851781   0.8255369  0.78037898 0.88878622 0.8813377 ]
Mean Dice: 0.8456
Parotid Dice (L,R): 0.7804, 0.8888
Combined Score: 0.8423
LR: 0.000062


100%|██████████| 198/198 [06:01<00:00,  1.83s/it]



Epoch 26
Train Loss: 0.2027
Val Loss: 0.0747
Per Class Dice: [0.92208058 0.80240668 0.80843445 0.85459762 0.90587899 0.85830963]
Mean Dice: 0.8459
Parotid Dice (L,R): 0.8546, 0.9059
Combined Score: 0.8562
LR: 0.000055


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 27
Train Loss: 0.1889
Val Loss: 0.1226
Per Class Dice: [0.6863635  0.77477221 0.82732091 0.74611899 0.70976306 0.77155384]
Mean Dice: 0.7659
Parotid Dice (L,R): 0.7461, 0.7098
Combined Score: 0.7545
LR: 0.000048


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 28
Train Loss: 0.2022
Val Loss: 0.0665
Per Class Dice: [0.81562601 0.79579296 0.7195493  0.7696165  0.91426208 0.77614946]
Mean Dice: 0.7951
Parotid Dice (L,R): 0.7696, 0.9143
Combined Score: 0.8091
LR: 0.000041


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 29
Train Loss: 0.1830
Val Loss: 0.0618
Per Class Dice: [0.8222177  0.88132736 0.84038131 0.83393154 0.89853381 0.90289062]
Mean Dice: 0.8714
Parotid Dice (L,R): 0.8339, 0.8985
Combined Score: 0.8699
LR: 0.000035


100%|██████████| 198/198 [05:55<00:00,  1.80s/it]



Epoch 30
Train Loss: 0.1957
Val Loss: 0.0558
Per Class Dice: [0.70049717 0.90085466 0.89483853 0.80596842 0.82834866 0.81866176]
Mean Dice: 0.8497
Parotid Dice (L,R): 0.8060, 0.8283
Combined Score: 0.8400
LR: 0.000029


100%|██████████| 198/198 [06:03<00:00,  1.84s/it]



Epoch 31
Train Loss: 0.1886
Val Loss: 0.0622
Per Class Dice: [0.82789636 0.89080419 0.86416679 0.8661972  0.81294337 0.88788478]
Mean Dice: 0.8644
Parotid Dice (L,R): 0.8662, 0.8129
Combined Score: 0.8570
LR: 0.000024


100%|██████████| 198/198 [06:01<00:00,  1.83s/it]



Epoch 32
Train Loss: 0.1847
Val Loss: 0.1023
Per Class Dice: [0.74861593 0.8640396  0.83567071 0.77864032 0.75022111 0.91284321]
Mean Dice: 0.8283
Parotid Dice (L,R): 0.7786, 0.7502
Combined Score: 0.8091
LR: 0.000019


100%|██████████| 198/198 [06:00<00:00,  1.82s/it]



Epoch 33
Train Loss: 0.1913
Val Loss: 0.0647
Per Class Dice: [0.81811919 0.84181309 0.85052578 0.87297684 0.87708204 0.86695007]
Mean Dice: 0.8619
Parotid Dice (L,R): 0.8730, 0.8771
Combined Score: 0.8658
LR: 0.000015


100%|██████████| 198/198 [06:02<00:00,  1.83s/it]



Epoch 34
Train Loss: 0.1761
Val Loss: 0.0699
Per Class Dice: [0.83830114 0.87103609 0.72627316 0.71425646 0.75828005 0.66691635]
Mean Dice: 0.7474
Parotid Dice (L,R): 0.7143, 0.7583
Combined Score: 0.7440
LR: 0.000011


100%|██████████| 198/198 [05:38<00:00,  1.71s/it]



Epoch 35
Train Loss: 0.1813
Val Loss: 0.0414
Per Class Dice: [0.84177604 0.90309919 0.77149504 0.95724489 0.91088137 0.92409986]
Mean Dice: 0.8934
Parotid Dice (L,R): 0.9572, 0.9109
Combined Score: 0.9056
Saved Best Parotid Model
LR: 0.000008


100%|██████████| 198/198 [05:57<00:00,  1.81s/it]



Epoch 36
Train Loss: 0.1796
Val Loss: 0.0577
Per Class Dice: [0.92746513 0.84773052 0.83802946 0.90942487 0.89592635 0.90345074]
Mean Dice: 0.8789
Parotid Dice (L,R): 0.9094, 0.8959
Combined Score: 0.8860
LR: 0.000005


100%|██████████| 198/198 [06:02<00:00,  1.83s/it]



Epoch 37
Train Loss: 0.1831
Val Loss: 0.0602
Per Class Dice: [0.73979389 0.92487469 0.84022282 0.82727747 0.98290718 0.82549316]
Mean Dice: 0.8802
Parotid Dice (L,R): 0.8273, 0.9829
Combined Score: 0.8876
LR: 0.000003


100%|██████████| 198/198 [05:53<00:00,  1.79s/it]



Epoch 38
Train Loss: 0.1919
Val Loss: 0.0837
Per Class Dice: [0.82181909 0.85948243 0.88173397 0.80679364 0.89916029 0.89007382]
Mean Dice: 0.8674
Parotid Dice (L,R): 0.8068, 0.8992
Combined Score: 0.8631
LR: 0.000001


100%|██████████| 198/198 [05:57<00:00,  1.81s/it]



Epoch 39
Train Loss: 0.1810
Val Loss: 0.0657
Per Class Dice: [0.82569826 0.72200863 0.67997596 0.71733968 0.7580368  0.89089301]
Mean Dice: 0.7537
Parotid Dice (L,R): 0.7173, 0.7580
Combined Score: 0.7489
LR: 0.000000


In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = nnUNet3D_CBAM(
    in_channels=1,
    out_channels=7,
    base_filters=24   # or try 32
).to(device)

model_path = "../experiments/custom_loss_cbam/04_rw_best_combined_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


C:\Users\sajju\AppData\Local\Temp\ipykernel_8864\2982874843.py:10: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


Model loaded for testing.


In [18]:
mean_dice, std_dice = evaluate_full_volume_cbam(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Volume: 272x357x357 | Patches: 245 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.927073467750758), np.float64(0.8250788644912628), np.float64(0.8397909536136404), np.float64(0.8424947147542339), np.float64(0.8215906849994171), np.float64(0.8787785708051714)]
Volume: 243x383x383 | Patches: 245 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.9235382309074847), np.float64(0.7611491567052371), np.float64(0.7296726509509457), np.float64(0.8274063236672776), np.float64(0.8328539470547434), np.float64(0.8970268930438825)]
Volume: 264x333x333 | Patches: 180 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.9099936261162758), np.float64(0.8315108579766588), np.float64(0.8270083787957806), np.float64(0.832150943522905), np.float64(0.8325785137481784), np.float64(0.8857138148867497)]
Volume: 248x395x395 | Patch

### spinal favored patching 

In [7]:
model = nnUNet3D_CBAM(
    in_channels=1,
    out_channels=7,
    base_filters=24   # or try 32
).to(device)

print("cbam nnU-Net style model Initialized")


cbam nnU-Net style model Initialized


In [8]:
EPOCHS = 40

optimizer = torch.optim.AdamW(model.parameters(), lr=2e-4)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=EPOCHS
)

weights = torch.tensor([0.05, 1, 1.3, 1.8, 2, 2, 1]).to(device)

loss_fn = InverseDistanceBoundaryDiceLoss(
    epsilon=1e-5,
    lambda_weight=0.4,
    class_weights = weights
)

scaler = torch.cuda.amp.GradScaler()

C:\Users\sajju\AppData\Local\Temp\ipykernel_15072\3779698161.py:16: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [12]:
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "custom_loss_cbam")

os.makedirs(SAVE_DIR, exist_ok=True)

best_val_loss = float("inf")
best_dice = 0.0
best_parotid = 0.0
best_combined = 0.0

for epoch in range(EPOCHS):

    # ------------------ TRAIN ------------------
    train_loss = train_one_epoch_cbam(
        model,
        train_loader,
        optimizer,
        scaler,
        loss_fn,
        device
    )

    # ------------------ VALIDATE ------------------
    val_loss, val_dice = validate_one_epoch_cbam(
        model,
        val_loader,
        loss_fn,
        device
    )

    # Ignore background (class 0)
    mean_dice = val_dice[1:].mean()

    # Parotid (L=4, R=5)
    parotid_score = (val_dice[3] + val_dice[4]) / 2
    # 🔥 Combined metric (IMPORTANT)
    combined_score = 0.7 * mean_dice + 0.3 * parotid_score

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")
    print(f"Mean Dice: {mean_dice:.4f}")
    print(f"Parotid Dice (L,R): {val_dice[3]:.4f}, {val_dice[4]:.4f}")
    print(f"Combined Score: {combined_score:.4f}")

    # ------------------ SAVE BEST LOSS ------------------
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "04_sc_best_loss_model.pth")
        )
        print("Saved Best Loss Model")

    # ------------------ SAVE BEST DICE ------------------
    if mean_dice > best_dice:
        best_dice = mean_dice
        save_checkpoint(
            model,
            optimizer,
            epoch,
            mean_dice,
            os.path.join(SAVE_DIR, "04_sc_best_dice_model.pth")
        )
        print("Saved Best Dice Model")

    # ------------------ SAVE BEST PAROTID ------------------
    if parotid_score > best_parotid:
        best_parotid = parotid_score
        save_checkpoint(
            model,
            optimizer,
            epoch,
            parotid_score,
            os.path.join(SAVE_DIR, "04_sc_best_parotid_model.pth")
        )
        print("Saved Best Parotid Model")

    # ------------------ SAVE BEST COMBINED ------------------
    if combined_score > best_combined:
        best_combined = combined_score
        save_checkpoint(
            model,
            optimizer,
            epoch,
            combined_score,
            os.path.join(SAVE_DIR, "04_sc_best_combined_model.pth")
        )
        print("Saved Best Combined Model")

    # ------------------ SAVE LAST MODEL ------------------
    save_checkpoint(
        model,
        optimizer,
        epoch,
        val_loss,
        os.path.join(SAVE_DIR, "04_sc_last_model.pth")
    )

    # ------------------ LR PRINT ------------------
    lr = optimizer.param_groups[0]['lr']
    print(f"LR: {lr:.6f}")

    # ------------------ SCHEDULER ------------------
    scheduler.step()

100%|██████████| 198/198 [05:45<00:00,  1.75s/it]



Epoch 0
Train Loss: 0.6713
Val Loss: 0.2481
Per Class Dice: [0.55383511 0.32923467 0.1764637  0.20525607 0.22222446 0.61146775]
Mean Dice: 0.3089
Parotid Dice (L,R): 0.2053, 0.2222
Combined Score: 0.2804
Saved Best Loss Model
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000200


100%|██████████| 198/198 [05:53<00:00,  1.79s/it]



Epoch 1
Train Loss: 0.3525
Val Loss: 0.0729
Per Class Dice: [0.64266226 0.55695556 0.73442239 0.40659277 0.52638326 0.76641407]
Mean Dice: 0.5982
Parotid Dice (L,R): 0.4066, 0.5264
Combined Score: 0.5587
Saved Best Loss Model
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000200


100%|██████████| 198/198 [05:58<00:00,  1.81s/it]



Epoch 2
Train Loss: 0.2947
Val Loss: 0.1236
Per Class Dice: [0.78772637 0.91002077 0.88213766 0.83081485 0.68396804 0.77316219]
Mean Dice: 0.8160
Parotid Dice (L,R): 0.8308, 0.6840
Combined Score: 0.7984
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000199


100%|██████████| 198/198 [06:02<00:00,  1.83s/it]



Epoch 3
Train Loss: 0.2639
Val Loss: 0.0869
Per Class Dice: [0.6963881  0.64642414 0.71167897 0.56518327 0.55655836 0.80712201]
Mean Dice: 0.6574
Parotid Dice (L,R): 0.5652, 0.5566
Combined Score: 0.6284
LR: 0.000197


100%|██████████| 198/198 [05:55<00:00,  1.79s/it]



Epoch 4
Train Loss: 0.2538
Val Loss: 0.0719
Per Class Dice: [0.92229429 0.8121886  0.80862824 0.61008566 0.86942469 0.89401265]
Mean Dice: 0.7989
Parotid Dice (L,R): 0.6101, 0.8694
Combined Score: 0.7811
Saved Best Loss Model
LR: 0.000195


100%|██████████| 198/198 [05:43<00:00,  1.74s/it]



Epoch 5
Train Loss: 0.2535
Val Loss: 0.0957
Per Class Dice: [0.81579226 0.73399629 0.82063358 0.70047288 0.72576842 0.85444762]
Mean Dice: 0.7671
Parotid Dice (L,R): 0.7005, 0.7258
Combined Score: 0.7509
LR: 0.000192


100%|██████████| 198/198 [06:19<00:00,  1.92s/it]



Epoch 6
Train Loss: 0.2426
Val Loss: 0.1782
Per Class Dice: [0.60667824 0.69401254 0.72038109 0.44696579 0.68777916 0.66772434]
Mean Dice: 0.6434
Parotid Dice (L,R): 0.4470, 0.6878
Combined Score: 0.6206
LR: 0.000189


100%|██████████| 198/198 [08:13<00:00,  2.49s/it]



Epoch 7
Train Loss: 0.2262
Val Loss: 0.1067
Per Class Dice: [0.82313982 0.71825498 0.72545182 0.66436821 0.87268125 0.65562371]
Mean Dice: 0.7273
Parotid Dice (L,R): 0.6644, 0.8727
Combined Score: 0.7397
Saved Best Parotid Model
LR: 0.000185


100%|██████████| 198/198 [06:47<00:00,  2.06s/it]



Epoch 8
Train Loss: 0.2168
Val Loss: 0.1172
Per Class Dice: [0.82470727 0.86913653 0.74746477 0.59369228 0.91274413 0.69608528]
Mean Dice: 0.7638
Parotid Dice (L,R): 0.5937, 0.9127
Combined Score: 0.7606
LR: 0.000181


100%|██████████| 198/198 [05:58<00:00,  1.81s/it]



Epoch 9
Train Loss: 0.2396
Val Loss: 0.0819
Per Class Dice: [0.72382811 0.79137424 0.78035426 0.68234318 0.75542154 0.81882802]
Mean Dice: 0.7657
Parotid Dice (L,R): 0.6823, 0.7554
Combined Score: 0.7516
LR: 0.000176


100%|██████████| 198/198 [06:03<00:00,  1.84s/it]



Epoch 10
Train Loss: 0.2277
Val Loss: 0.0718
Per Class Dice: [0.92804043 0.74532109 0.8385045  0.83048329 0.88940134 0.88933425]
Mean Dice: 0.8386
Parotid Dice (L,R): 0.8305, 0.8894
Combined Score: 0.8450
Saved Best Loss Model
Saved Best Dice Model
Saved Best Parotid Model
Saved Best Combined Model
LR: 0.000171


100%|██████████| 198/198 [05:50<00:00,  1.77s/it]



Epoch 11
Train Loss: 0.2071
Val Loss: 0.0553
Per Class Dice: [0.94460601 0.90536732 0.89453214 0.75505564 0.84683697 0.82220599]
Mean Dice: 0.8448
Parotid Dice (L,R): 0.7551, 0.8468
Combined Score: 0.8316
Saved Best Loss Model
Saved Best Dice Model
LR: 0.000165


100%|██████████| 198/198 [06:09<00:00,  1.87s/it]



Epoch 12
Train Loss: 0.2310
Val Loss: 0.0695
Per Class Dice: [0.70428621 0.88737041 0.71671683 0.5349648  0.69670984 0.69666906]
Mean Dice: 0.7065
Parotid Dice (L,R): 0.5350, 0.6967
Combined Score: 0.6793
LR: 0.000159


100%|██████████| 198/198 [06:25<00:00,  1.95s/it]



Epoch 13
Train Loss: 0.2045
Val Loss: 0.0701
Per Class Dice: [0.83409159 0.76723732 0.79961901 0.77211919 0.75347654 0.78963382]
Mean Dice: 0.7764
Parotid Dice (L,R): 0.7721, 0.7535
Combined Score: 0.7723
LR: 0.000152


100%|██████████| 198/198 [06:16<00:00,  1.90s/it]



Epoch 14
Train Loss: 0.2109
Val Loss: 0.1162
Per Class Dice: [0.80714147 0.8725901  0.76366549 0.79087894 0.82870666 0.77042337]
Mean Dice: 0.8053
Parotid Dice (L,R): 0.7909, 0.8287
Combined Score: 0.8066
LR: 0.000145


100%|██████████| 198/198 [05:59<00:00,  1.82s/it]



Epoch 15
Train Loss: 0.2012
Val Loss: 0.0824
Per Class Dice: [0.70519626 0.87817812 0.7522596  0.67205155 0.67944285 0.60185401]
Mean Dice: 0.7168
Parotid Dice (L,R): 0.6721, 0.6794
Combined Score: 0.7045
LR: 0.000138


100%|██████████| 198/198 [06:08<00:00,  1.86s/it]



Epoch 16
Train Loss: 0.2064
Val Loss: 0.0727
Per Class Dice: [0.74350343 0.82114734 0.82733625 0.52139754 0.73540629 0.90396102]
Mean Dice: 0.7618
Parotid Dice (L,R): 0.5214, 0.7354
Combined Score: 0.7218
LR: 0.000131


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 17
Train Loss: 0.1987
Val Loss: 0.1513
Per Class Dice: [0.71736195 0.86329193 0.63279077 0.63916137 0.7687082  0.87978477]
Mean Dice: 0.7567
Parotid Dice (L,R): 0.6392, 0.7687
Combined Score: 0.7409
LR: 0.000123


100%|██████████| 198/198 [06:12<00:00,  1.88s/it]



Epoch 18
Train Loss: 0.1989
Val Loss: 0.0605
Per Class Dice: [0.94374673 0.86569254 0.74135791 0.87133508 0.75540141 0.90727815]
Mean Dice: 0.8282
Parotid Dice (L,R): 0.8713, 0.7554
Combined Score: 0.8238
LR: 0.000116


100%|██████████| 198/198 [06:04<00:00,  1.84s/it]



Epoch 19
Train Loss: 0.1952
Val Loss: 0.0532
Per Class Dice: [0.70834255 0.79914103 0.83884329 0.83664107 0.74179663 0.90608667]
Mean Dice: 0.8245
Parotid Dice (L,R): 0.8366, 0.7418
Combined Score: 0.8139
Saved Best Loss Model
LR: 0.000108


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 20
Train Loss: 0.2040
Val Loss: 0.0746
Per Class Dice: [0.825164   0.87286458 0.72334053 0.82986409 0.83110762 0.67481362]
Mean Dice: 0.7864
Parotid Dice (L,R): 0.8299, 0.8311
Combined Score: 0.7996
LR: 0.000100


100%|██████████| 198/198 [06:41<00:00,  2.03s/it]



Epoch 21
Train Loss: 0.1852
Val Loss: 0.0690
Per Class Dice: [0.91588132 0.8090823  0.81439683 0.82377388 0.75995407 0.87294344]
Mean Dice: 0.8160
Parotid Dice (L,R): 0.8238, 0.7600
Combined Score: 0.8088
LR: 0.000092


100%|██████████| 198/198 [07:04<00:00,  2.15s/it]



Epoch 22
Train Loss: 0.1842
Val Loss: 0.0781
Per Class Dice: [0.91488999 0.7001478  0.83160028 0.86220556 0.81274117 0.85150489]
Mean Dice: 0.8116
Parotid Dice (L,R): 0.8622, 0.8127
Combined Score: 0.8194
LR: 0.000084


100%|██████████| 198/198 [07:04<00:00,  2.15s/it]



Epoch 23
Train Loss: 0.1925
Val Loss: 0.0787
Per Class Dice: [0.6228419  0.79988575 0.87726098 0.70581719 0.8184692  0.69337555]
Mean Dice: 0.7790
Parotid Dice (L,R): 0.7058, 0.8185
Combined Score: 0.7739
LR: 0.000077


100%|██████████| 198/198 [06:14<00:00,  1.89s/it]



Epoch 24
Train Loss: 0.1897
Val Loss: 0.0496
Per Class Dice: [0.96530758 0.90937424 0.86392892 0.88712802 0.81980263 0.83334636]
Mean Dice: 0.8627
Parotid Dice (L,R): 0.8871, 0.8198
Combined Score: 0.8599
Saved Best Loss Model
Saved Best Dice Model
Saved Best Combined Model
LR: 0.000069


100%|██████████| 198/198 [06:33<00:00,  1.99s/it]



Epoch 25
Train Loss: 0.1776
Val Loss: 0.0685
Per Class Dice: [0.80637848 0.9430626  0.92055213 0.77673753 0.86125691 0.91761806]
Mean Dice: 0.8838
Parotid Dice (L,R): 0.7767, 0.8613
Combined Score: 0.8644
Saved Best Dice Model
Saved Best Combined Model
LR: 0.000062


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 26
Train Loss: 0.1854
Val Loss: 0.0900
Per Class Dice: [0.62161661 0.80244974 0.78411553 0.83032843 0.57066879 0.81225694]
Mean Dice: 0.7600
Parotid Dice (L,R): 0.8303, 0.5707
Combined Score: 0.7421
LR: 0.000055


100%|██████████| 198/198 [06:17<00:00,  1.91s/it]



Epoch 27
Train Loss: 0.1786
Val Loss: 0.0823
Per Class Dice: [0.92361095 0.85410236 0.77864597 0.63982395 0.78392089 0.85334353]
Mean Dice: 0.7820
Parotid Dice (L,R): 0.6398, 0.7839
Combined Score: 0.7609
LR: 0.000048


100%|██████████| 198/198 [06:16<00:00,  1.90s/it]



Epoch 28
Train Loss: 0.1732
Val Loss: 0.1052
Per Class Dice: [0.82502477 0.84190609 0.84754736 0.69491854 0.86808008 0.91194284]
Mean Dice: 0.8329
Parotid Dice (L,R): 0.6949, 0.8681
Combined Score: 0.8175
LR: 0.000041


100%|██████████| 198/198 [05:49<00:00,  1.77s/it]



Epoch 29
Train Loss: 0.1948
Val Loss: 0.0574
Per Class Dice: [0.93308416 0.74717368 0.74297592 0.89943608 0.87897062 0.79508344]
Mean Dice: 0.8127
Parotid Dice (L,R): 0.8994, 0.8790
Combined Score: 0.8357
Saved Best Parotid Model
LR: 0.000035


100%|██████████| 198/198 [06:09<00:00,  1.86s/it]



Epoch 30
Train Loss: 0.1799
Val Loss: 0.0895
Per Class Dice: [0.87048128 0.72848252 0.75027509 0.73397343 0.73686043 0.75582765]
Mean Dice: 0.7411
Parotid Dice (L,R): 0.7340, 0.7369
Combined Score: 0.7394
LR: 0.000029


100%|██████████| 198/198 [06:01<00:00,  1.82s/it]



Epoch 31
Train Loss: 0.1842
Val Loss: 0.0990
Per Class Dice: [0.91312105 0.86963181 0.85775269 0.67745422 0.68064715 0.82384155]
Mean Dice: 0.7819
Parotid Dice (L,R): 0.6775, 0.6806
Combined Score: 0.7510
LR: 0.000024


100%|██████████| 198/198 [05:59<00:00,  1.81s/it]



Epoch 32
Train Loss: 0.1781
Val Loss: 0.0535
Per Class Dice: [0.96762447 0.91750126 0.84438615 0.75690434 0.7976883  0.82599387]
Mean Dice: 0.8285
Parotid Dice (L,R): 0.7569, 0.7977
Combined Score: 0.8131
LR: 0.000019


100%|██████████| 198/198 [05:56<00:00,  1.80s/it]



Epoch 33
Train Loss: 0.1786
Val Loss: 0.0562
Per Class Dice: [0.82006382 0.88554774 0.86652878 0.9296264  0.6913124  0.77542187]
Mean Dice: 0.8297
Parotid Dice (L,R): 0.9296, 0.6913
Combined Score: 0.8239
LR: 0.000015


100%|██████████| 198/198 [05:54<00:00,  1.79s/it]



Epoch 34
Train Loss: 0.1898
Val Loss: 0.0961
Per Class Dice: [0.8862407  0.79006563 0.87669114 0.8924959  0.6741339  0.71029123]
Mean Dice: 0.7887
Parotid Dice (L,R): 0.8925, 0.6741
Combined Score: 0.7871
LR: 0.000011


100%|██████████| 198/198 [06:07<00:00,  1.86s/it]



Epoch 35
Train Loss: 0.1754
Val Loss: 0.0704
Per Class Dice: [0.91719    0.88754105 0.8362414  0.75084618 0.87701811 0.8828804 ]
Mean Dice: 0.8469
Parotid Dice (L,R): 0.7508, 0.8770
Combined Score: 0.8370
LR: 0.000008


100%|██████████| 198/198 [06:03<00:00,  1.83s/it]



Epoch 36
Train Loss: 0.1843
Val Loss: 0.0558
Per Class Dice: [0.82848875 0.83051995 0.88081627 0.87856166 0.89492357 0.78941266]
Mean Dice: 0.8548
Parotid Dice (L,R): 0.8786, 0.8949
Combined Score: 0.8644
Saved Best Combined Model
LR: 0.000005


100%|██████████| 198/198 [05:58<00:00,  1.81s/it]



Epoch 37
Train Loss: 0.1764
Val Loss: 0.0751
Per Class Dice: [0.81260267 0.75480664 0.8496281  0.64593192 0.79731268 0.89692265]
Mean Dice: 0.7889
Parotid Dice (L,R): 0.6459, 0.7973
Combined Score: 0.7687
LR: 0.000003


100%|██████████| 198/198 [05:44<00:00,  1.74s/it]



Epoch 38
Train Loss: 0.1679
Val Loss: 0.0589
Per Class Dice: [0.95848953 0.69306105 0.8858419  0.83236391 0.86825248 0.90110632]
Mean Dice: 0.8361
Parotid Dice (L,R): 0.8324, 0.8683
Combined Score: 0.8404
LR: 0.000001


100%|██████████| 198/198 [05:51<00:00,  1.78s/it]



Epoch 39
Train Loss: 0.1709
Val Loss: 0.0750
Per Class Dice: [0.82814556 0.76275625 0.79127906 0.74265079 0.68364233 0.76492885]
Mean Dice: 0.7491
Parotid Dice (L,R): 0.7427, 0.6836
Combined Score: 0.7383
LR: 0.000000


In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = nnUNet3D_CBAM(
    in_channels=1,
    out_channels=7,
    base_filters=24
).to(device)

model_path = "../experiments/custom_loss_cbam/04_sc_best_combined_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


C:\Users\sajju\AppData\Local\Temp\ipykernel_15072\2547328426.py:10: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


Model loaded for testing.


In [14]:
mean_dice, std_dice = evaluate_full_volume_cbam(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Volume: 272x357x357 | Patches: 245 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.9338378264927643), np.float64(0.8561212593423077), np.float64(0.8182610214004861), np.float64(0.8376716541938187), np.float64(0.8402318724293381), np.float64(0.9089318291989776)]
Volume: 243x383x383 | Patches: 245 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.9145614035337539), np.float64(0.67306865438364), np.float64(0.7346938780098687), np.float64(0.8214687852987184), np.float64(0.8317375208529698), np.float64(0.898038055133256)]
Volume: 264x333x333 | Patches: 180 | Stride: 48
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.9018812009077821), np.float64(0.8415841585328918), np.float64(0.8136328430462698), np.float64(0.8267786125889095), np.float64(0.8149386846512672), np.float64(0.902296410037844)]
Volume: 248x395x395 | Patches

### Results

### Model Evaluation Comparison

### lambda = 0.4

| Model & Configuration | Class 1 | Class 2 | Class 3 | Class 4 (Parotid L) | Class 5 (Parotid R) | Class 6 | Mean Dice |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **Model 1 (Initial Weights)**<br>`[0.05, 1, 1, 1.2, 1.8, 1.8, 1]` | 0.8960 | 0.8017 | 0.7107 | 0.8102 | 0.7981 | 0.8434 | 0.8100 |
| **Model 2 (Revised Weights)**<br>`[0.05, 1, 1.2, 1.5, 2, 2, 1]` | **0.9132** | **0.8017** | **0.7583** | 0.8170 | 0.7837 | 0.8353 | **0.8182** |
| **Model 3 (Spinal Favored Patching)**<br>`[0.05, 1, 1.3, 1.8, 2, 2, 1]` | 0.9119 | 0.7586 | 0.7461 | **0.8180** | **0.8033** | **0.8556** | 0.8156 |

**Observations**:
* **Model 2 (Revised Weights)** achieved the highest overall **Mean Dice (0.8182)** and performed best on Classes 1, 2, and 3.
* **Model 3 (Spinal Favored Patching)** sacrificed some performance on Class 2, but achieved the best results for the Parotid glands (Classes 4 & 5) and Class 6.


### Model Evaluation Comparison (Lambda = 0.6)

| Model & Configuration | Class 1 | Class 2 | Class 3 | Class 4 (Parotid L) | Class 5 (Parotid R) | Class 6 | Mean Dice |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **Model 1**<br>`[0.05, 1, 1, 1.2, 1.8, 1.8, 1]` | 0.9047 ± 0.0235 | **0.8100 ± 0.0349** | 0.7394 ± 0.0585 | 0.7768 ± 0.0279 | 0.7894 ± 0.0770 | 0.8007 ± 0.0609 | 0.8035 |
| **Model 2**<br>`[0.05, 1, 1.2, 1.5, 2, 2, 1]` | **0.9142 ± 0.0266** | 0.8083 ± 0.0324 | **0.7553 ± 0.0526** | **0.8094 ± 0.0298** | **0.7975 ± 0.0571** | **0.8559 ± 0.0362** | **0.8234** |

**Observations (λ=0.6)**:
* **Model 2** significantly outperforms Model 1 across almost all classes, achieving a higher overall Mean Dice (0.8234).
* The Parotid glands (Classes 4 & 5) and Class 6 show noticeable improvements in Model 2 upon revising the class weights.
